packages/jest-environment-puppeteer/src/PuppeteerEnvironment.js

```diff
@@ -115,16 +115,25 @@ class PuppeteerEnvironment extends NodeEnvironment {
   }
 
   async teardown() {
-    const config = await readConfig()
-    this.global.page.removeListener('pageerror', handleError)
+    const { page, context, browser, puppeteerConfig } = this.global
 
-    if (config.browserContext === 'incognito') {
-      await this.global.context.close()
+    if (page) {
+      page.removeListener('pageerror', handleError)
+    }
+
+    if (puppeteerConfig.browserContext === 'incognito') {
+      if (context) {
+        await context.close()
+      }
     } else {
-      await this.global.page.close()
```

Comment:
tbh logic would stay the same if you will change lines L120-L132 to:
```
if (puppeteerConfig.browserContext === 'incognito' && context) {
  await context.close()
} else if (page) {
  page.removeListener('pageerror', handleError)
  await page.close()
}
```

Reply:
This code is different: If browserContent === 'incognito' && context, then page.removeListener, wont be called, but that call should only depend on existance of page.

```
if (puppeteerConfig.browserContext === 'incognito' && context) {
  await context.close()
} else if (page) {
  await page.close()
}
```
Normally this also changes the logic of the original block i posted, but since page and context depend on each other, so when browserContent === 'incognito' && context === undefined (because error in setup), it must be page === undefined, because there cant be a page without context.

Reply:
To extend on this: You must have knowledge about the dependency between context & page, to correctly understand the minified if/else you propose, which makes it harder to reason about what happens.